### PR TITLE
man page for docker daemon command

### PR DESCRIPTION
**Pull request by shishir-a412ed**
Signed-off-by: Shishir Mahajan shishir.mahajan@redhat.com

### CODE REVIEW

**Comment by shishir-a412ed** [comment]
ping @tiborvass

File: man/docker.1.md [hunk 1]
```diff
@@ -7,24 +7,29 @@ docker \- Docker image and container command line interface
...
+and will be removed soon.
 
-The Docker CLI has over 30 commands. The commands are listed below and each has
-its own man page which explain usage and arguments.
+You can check the daemon flags using **docker daemon --help**.
+Daemon flags should be specified after the **daemon** keyword in the following
+format.
 
-To see the man page for a command run **man docker <command>**.
+**docker [OPTIONS] daemon [arg...]**
```
Comment (tiborvass) [inline man/docker.1.md]: It's `docker daemon [OPTIONS]`

File: man/docker.1.md [hunk 2]
```diff
@@ -7,24 +7,29 @@ docker \- Docker image and container command line interface
...
 **docker** has two distinct functions. It is used for starting the Docker
 daemon and to run the CLI (i.e., to command the daemon to manage images,
 containers etc.) So **docker** is both a server, as a daemon, and a client
 to the daemon, through the CLI.
 
-To run the Docker daemon you do not specify any of the commands listed below but
-must specify the **-d** option.  The other options listed below are for the
-daemon only.
+To run the Docker daemon you can specify **docker daemon**. You can also
+start daemon using **-d** or **--daemon**. However these options are deprecated
```
Comment (tiborvass) [inline man/docker.1.md]: if they are deprecated, no need to mention them

File: man/docker.1.md (hunk 1 above)
Comment (shishir-a412ed) [inline man/docker.1.md]: Fixed.

File: man/docker.1.md (hunk 2 above)
Comment (shishir-a412ed) [inline man/docker.1.md]: Fixed.

**Comment by thaJeztah** [comment]
LGTM, thanks!

**Comment by SvenDowideit** [comment]
I kinda feel that there should be a separate `docker-daemon.1.md`. That way, a client only OS, like OSX could elide optionally elide that one, as its not usable on that platform.

**Comment by thaJeztah** [comment]
@SvenDowideit yes, I was thinking about that as well (didn't think of the "client-only" case, which is interesting). Do you want that change as part of this PR, or a separate one?

**Comment by shishir-a412ed** [comment]
@SvenDowideit @thaJeztah +1 for having a separate `docker-daemon.1.md`.
I can make the changes once we have consensus on this.

**Comment by SvenDowideit** [comment]
@icecrime @moxiegirl @tianon @ipbabble @sallyom what do you think?

**Comment by moxiegirl** [comment]
+1 @shishir-a412ed  I think it goes in its own man page category. I forget what it is... let me do a quick check.

I also think this content needs tweaking...

File: man/docker.1.md [hunk 3]
```diff
@@ -7,24 +7,26 @@ docker \- Docker image and container command line interface
 # SYNOPSIS
 **docker** [OPTIONS] COMMAND [arg...]
```
Comment (moxiegirl) [inline man/docker.1.md]: If this were to just document the cli, then following the example of standad `man` page, we would move all the daemon discussion into a `docker.8.md` file and point user to that man page. For example:

---> https://gist.github.com/moxiegirl/86730388df360b27ebfb

% DOCKER(1) Docker User Manuals
% William Henry
% APRIL 2014
# NAME
docker \- Docker command line interface

# SYNOPSIS
**docker** [OPTIONS] COMMAND [arg...]

**docker** daemon [ --help | ... ]

**docker** [ --help | -v | --version ]

# DESCRIPTION
Docker is a program that packs, ships, and runs applications as lightweight containers. The **docker** command runs a command-line client that manages images and conatiners. The **docker**(8) daemon runs in the background on your machine. The daemon receives commands from the client and serves the containers as instructed. 

To run the Docker daemon you can specify **docker daemon**.  You can check the daemon options using **docker daemon --help**. To see the **man** page for the daemon, run **man 8 docker**.

The Docker CLI has over 30 commands. The commands are listed below and each has
its own man page which explain usage and arguments.

To see the man page for a command run **man docker <command>**.

# GLOBAL OPTIONS
**--help**
  Print usage statement

**--config**=""
  Specifies the location of the Docker client configuration files. The default is '~/.docker'.

**-D**, **--debug**=*true*|*false*
  Enable debug mode. Default is false.

**-H**, **--host**=[unix:///var/run/docker.sock]: tcp://[host:port] to bind or
unix://[/path/to/socket] to use.
  The socket(s) to bind to in daemon mode specified using one or more
  tcp://host:port, unix:///path/to/socket, fd://* or fd://socketfd.

**-l**, **--log-level**="*debug*|*info*|*warn*|*error*|*fatal*""
  Set the logging level. Default is `info`.

**--tls**=*true*|*false*
  Use TLS; implied by --tlsverify. Default is false.

**--tlscacert**=~/.docker/ca.pem
  Trust certs signed only by this CA.

**--tlscert**=~/.docker/cert.pem
  Path to TLS certificate file.

**--tlskey**=~/.docker/key.pem
  Path to TLS key file.

**--tlsverify**=*true*|*false*
  Use TLS and verify the remote (daemon: verify client, client: verify daemon).
  Default is false.

**-v**, **--version**=*true*|*false*
  Print version information and quit. Default is false.

# COMMANDS
**attach**
  Attach to a running container
  See **docker-attach(1)** for full documentation on the **attach** command.

**build**
  Build an image from a Dockerfile
  See **docker-build(1)** for full documentation on the **build** command.

**commit**
  Create a new image from a container's changes
  See **docker-commit(1)** for full documentation on the **commit** command.

**cp**
  Copy files/folders from a container's filesystem to the host
  See **docker-cp(1)** for full documentation on the **cp** command.

**create**
  Create a new container
  See **docker-create(1)** for full documentation on the **create** command.

**diff**
  Inspect changes on a container's filesystem
  See **docker-diff(1)** for full documentation on the **diff** command.

**events**
  Get real time events from the server
  See **docker-events(1)** for full documentation on the **events** command.

**exec**
  Run a command in a running container
  See **docker-exec(1)** for full documentation on the **exec** command.

**export**
  Stream the contents of a container as a tar archive
  See **docker-export(1)** for full documentation on the **export** command.

**history**
  Show the history of an image
  See **docker-history(1)** for full documentation on the **history** command.

**images**
  List images
  See **docker-images(1)** for full documentation on the **images** command.

**import**
  Create a new filesystem image from the contents of a tarball
  See **docker-import(1)** for full documentation on the **import** command.

**info**
  Display system-wide information
  See **docker-info(1)** for full documentation on the **info** command.

**inspect**
  Return low-level information on a container or image
  See **docker-inspect(1)** for full documentation on the **inspect** command.

**kill**
  Kill a running container (which includes the wrapper process and everything
inside it)
  See **docker-kill(1)** for full documentation on the **kill** command.

**load**
  Load an image from a tar archive
  See **docker-load(1)** for full documentation on the **load** command.

**login**
  Register or login to a Docker Registry
  See **docker-login(1)** for full documentation on the **login** command.

**logout**
  Log the user out of a Docker Registry
  See **docker-logout(1)** for full documentation on the **logout** command.

**logs**
  Fetch the logs of a container
  See **docker-logs(1)** for full documentation on the **logs** command.

**pause**
  Pause all processes within a container
  See **docker-pause(1)** for full documentation on the **pause** command.

**port**
  Lookup the public-facing port which is NAT-ed to PRIVATE_PORT
  See **docker-port(1)** for full documentation on the **port** command.

**ps**
  List containers
  See **docker-ps(1)** for full documentation on the **ps** command.

**pull**
  Pull an image or a repository from a Docker Registry
  See **docker-pull(1)** for full documentation on the **pull** command.

**push**
  Push an image or a repository to a Docker Registry
  See **docker-push(1)** for full documentation on the **push** command.

**rename**
  Rename a container.
  See **docker-rename(1)** for full documentation on the **rename** command.

**restart**
  Restart a running container
  See **docker-restart(1)** for full documentation on the **restart** command.

**rm**
  Remove one or more containers
  See **docker-rm(1)** for full documentation on the **rm** command.

**rmi**
  Remove one or more images
  See **docker-rmi(1)** for full documentation on the **rmi** command.

**run**
  Run a command in a new container
  See **docker-run(1)** for full documentation on the **run** command.

**save**
  Save an image to a tar archive
  See **docker-save(1)** for full documentation on the **save** command.

**search**
  Search for an image in the Docker index
  See **docker-search(1)** for full documentation on the **search** command.

**start**
  Start a stopped container
  See **docker-start(1)** for full documentation on the **start** command.

**stats**
  Display a live stream of one or more containers' resource usage statistics
  See **docker-stats(1)** for full documentation on the **stats** command.

**stop**
  Stop a running container
  See **docker-stop(1)** for full documentation on the **stop** command.

**tag**
  Tag an image into a repository
  See **docker-tag(1)** for full documentation on the **tag** command.

**top**
  Lookup the running processes of a container
  See **docker-top(1)** for full documentation on the **top** command.

**unpause**
  Unpause all processes within a container
  See **docker-unpause(1)** for full documentation on the **unpause** command.

**version**
  Show the Docker version information
  See **docker-version(1)** for full documentation on the **version** command.

**wait**
  Block until a container stops, then print its exit code
  See **docker-wait(1)** for full documentation on the **wait** command.

# HISTORY
April 2014, Originally compiled by William Henry (whenry at redhat dot com) based on docker.com source material and internal work.

**Comment by shishir-a412ed** [comment]
@moxiegirl Can we have `docker-daemon.8.md` instead of `docker.8.md`

And in `docker.1.md`, we would have a reference something like:
'See **docker-daemon(8)** for full documentation on the **docker daemon**.'

What do you think ?

**Comment by ipbabble** [comment]
Sorry for the late comment. Thanks for pulling me into this. 
So the man docker would tell you to run man docker daemon or man docker <command> ? I'm not sure having a "man docker daemon" is the standard Linux way of doing this.

**Comment by tianon** [comment]
I'm +1 for a separate "docker-daemon" man page.  ("man docker-daemon")

**Comment by ipbabble** [comment]
Hi above says "To run the Docker daemon you can specify docker daemon". Has this been done? I missed that. So "docker -d" is depreciated for "docker daemon"? If so, then yes, this makes much more sense. +1.

**Comment by shishir-a412ed** [comment]
@ipbabble yes docker 1.8 onwards `docker -d` is deprecated. It is replaced with new `docker daemon` command.

**Comment by ipbabble** [comment]
So +1. Absolutely.

**Comment by moxiegirl** [comment]
@tianon @ipbabble  The standard way to get to the daemon docs is

`man 8 docker`  and this is mentioned in the suggested text of the `docker.1.md` page. 

Its been a while since I spent a lot of time writing `man` pages but they have their own traditions. I haven't seen any precedence for naming the man page differently than the actual object it documents --- this  is what you folks are suggesting with naming the file `docker-daemon.8.md` remember, the `man` page user doesn't see the file name just the command.

Do y'all have a reference for using the construction you are suggesting? It would help me understand your position. (BTW unfortunate that we didn't just name the thing `dockerd` --- which` dockerd.8.md` would to me be a better name (ssh/sshd)

**Comment by ipbabble** [comment]
@moxiegirl nope I don't think the 'man 8 docker' is standard. see other's like 'man docker run' etc.

**Comment by moxiegirl** [comment]
@ipbabble `docker run` is a subcommand --- I don't see where it is analogous.

**Comment by thaJeztah** [comment]
@moxiegirl `daemon` is now a sub command; ie `docker daemon --option-x` will start the docker daemon.

**Comment by moxiegirl** [comment]
@thaJeztah brilliant! I keep smacking my face into that fact.  At some point, my brain will get it.  `docker-daemon.8.md` makes sense then.  @shishir-a412ed & @ipbabble apologies for the churn while I got sorted.

@shishir-a412ed Let us know if you want to pick up this extra work or not.

**Comment by shishir-a412ed** [comment]
@moxiegirl I am on it :)

**Comment by rhatdan** [comment]
BTW You can do both.

man 8 docker 

Would bring up docker-daemon.8.1

You just need to install a reference.

cat docker.8
.so man8/docker-daemon.8

**Comment by shishir-a412ed** [comment]
@moxiegirl 
This is WIP. I ll update you once I am done, so that you can review.

Shishir

**Comment by moxiegirl** [comment]
@shishir-a412ed We should move the storage driver options into the `docker-daemon.8.md` as well -- I believe they all belong there.

**Comment by shishir-a412ed** [comment]
@moxiegirl 
ah ! I missed that one. Please check now.

**Comment by moxiegirl** [comment]
@shishir-a412ed I missed it you need to update the page HISTORY on both.  Before you do that, let's run it by @thaJeztah again as he always finds something and there have been a lot of changes since his original approval.  Apologies this is taking so long...

File: man/docker.1.md [hunk 4]
```diff
@@ -26,125 +30,20 @@ To see the man page for a command run **man docker <command>**.
...
-  Attach containers to a pre\-existing network bridge; use 'none' to disable container networking
-
-**--bip**=""
-  Use the provided CIDR notation address for the dynamically created bridge (docker0); Mutually exclusive of \-b
-
 **--config**=""
   Specifies the location of the Docker client configuration files. The default is '~/.docker'.
 
 **-D**, **--debug**=*true*|*false*
   Enable debug mode. Default is false.
```
Comment (thaJeztah) [inline man/docker.1.md]: (Sorry seems I cannot comment on a *deleted* line)

should we still keep the `-d, --daemon` around? It's still there for backward compatibility (but marked for removal in 1.10)

Comment (shishir-a412ed) [inline man/docker.1.md]: @thaJeztah does this help ?
https://github.com/docker/docker/pull/16063#discussion_r38711266

Comment (thaJeztah) [inline man/docker.1.md]: @shishir-a412ed possibly (although the `-d` option is not for the daemon command). 

I might have other comments coming up, but got something else pop-up when I started reviewing :blush:

**Comment by SvenDowideit** [comment]
LGTM :)

File: man/docker-daemon.8.md [hunk 5]
```diff
@@ -0,0 +1,402 @@
+% DOCKER(1) Docker User Manuals
```
Comment (thaJeztah) [inline man/docker-daemon.8.md]: Should this be `DOCKER DAEMON (8)`, or something along that line (not too well known with man conventions)

File: man/docker-daemon.8.md [hunk 6]
```diff
@@ -0,0 +1,402 @@
...
+
+% William Henry
```
Comment (thaJeztah) [inline man/docker-daemon.8.md]: If we're talking about the changes; not sure what name should be here (perhaps yours, given that it's a new file), idk :smile:

File: man/docker-daemon.8.md [hunk 7]
```diff
@@ -0,0 +1,402 @@
+% DOCKER(1) Docker User Manuals
+
+% William Henry 
+
+% SEPTEMBER 2015
+
+# NAME
+docker-daemon - Enable daemon mode
```
Comment (thaJeztah) [inline man/docker-daemon.8.md]: Just a suggestion (open to better alternatives);

    docker-daemon - Run (/ Start) Docker in daemon mode

File: man/docker-daemon.8.md [hunk 8]
```diff
@@ -0,0 +1,402 @@
...
+**-H**, **--host**=[unix:///var/run/docker.sock]: tcp://[host:port] to bind or
+unix://[/path/to/socket] to use.
+  The socket(s) to bind to in daemon mode specified using one or more
+  tcp://host:port, unix:///path/to/socket, fd://* or fd://socketfd.
+
+**--help**
+  Print usage statement
+
+**--icc**=*true*|*false*
+  Allow unrestricted inter\-container and Docker daemon host communication. If disabled, containers can still be linked together using **--link** option (see **docker-run(1)**). Default is true.
```
Comment (thaJeztah) [inline man/docker-daemon.8.md]: using *the* **--link** option (missing "the")

File: man/docker-daemon.8.md (hunk 6 above)
Comment (shishir-a412ed) [inline man/docker-daemon.8.md]: I ll be happy to put my name in :) .. I ll update the file.

Comment (shishir-a412ed) [inline man/docker-daemon.8.md]: Fixed.

File: man/docker-daemon.8.md (hunk 8 above)
Comment (shishir-a412ed) [inline man/docker-daemon.8.md]: Fixed.

**Comment by shishir-a412ed** [comment]
@moxiegirl @SvenDowideit 

https://github.com/docker/docker/pull/16063#discussion_r39661648
https://github.com/docker/docker/pull/16063#discussion_r39662376

WDYT ?

File: man/docker.1.md [hunk 9]
```diff
@@ -26,125 +30,20 @@ To see the man page for a command run **man docker <command>**.
...
-**-G**, **--group**=""
-  Group to assign the unix socket specified by -H when running in daemon mode.
-  use '' (the empty string) to disable setting of a group. Default is `docker`.
-
-**-g**, **--graph**=""
-  Path to use as the root of the Docker runtime. Default is `/var/lib/docker`.
-
 **-H**, **--host**=[unix:///var/run/docker.sock]: tcp://[host:port] to bind or
 unix://[/path/to/socket] to use.
   The socket(s) to bind to in daemon mode specified using one or more
```
Comment (thaJeztah) [inline man/docker.1.md]: should this part still be here, given that this describes the *daemon* address/socket to listen on?

File: man/docker-daemon.8.md (hunk 5 above)
Comment (thaJeztah) [inline man/docker-daemon.8.md]: hm GitHub hiccups - I put this comment on the `-H` flag in `docker.1.md`

Comment (thaJeztah) [inline man/docker-daemon.8.md]: umf, now it's back in the right location, pls ignore (I'm on mobile) :'(

Comment (shishir-a412ed) [inline man/docker-daemon.8.md]: no worries :)

**Comment by thaJeztah** [comment]
Thanks! couple of suggestions and nits, but looking good!

File: man/docker-daemon.8.md (hunk 7 above)
Comment (shishir-a412ed) [inline man/docker-daemon.8.md]: This (`Enable daemon mode`) is inline with what `docker daemon --help` displays. Let's wait for other maintainers to comment.

**Comment by thaJeztah** [comment]
Just a heads up; PR https://github.com/docker/docker/pull/16031 is about to be merged, and adds a new `--dns-opt` option.

**Comment by tiborvass** [comment]
Ping @thaJeztah @moxiegirl we need this

**Comment by moxiegirl** [comment]
LGTM

**Comment by thaJeztah** [comment]
Thanks for the ping @tiborvass, I think there are three outstanding comments, perhaps @moxiegirl can have a look at those, but "LGTM" apart from those.

**Comment by thaJeztah** [comment]
Hmf, GitHub cache, LOL. Thanks @moxiegirl

**Comment by moxiegirl** [comment]
@thaJeztah I've put wrapping this up on my weekend todo list.  Thanks @shishir-a412ed for being patient.  Nicely done PR!

**Comment by thaJeztah** [comment]
@moxiegirl all good, thanks!

**Comment by thaJeztah** [comment]
And thanks @shishir-a412ed !

**Pull request merged**